### ISSUE TITLE
@clean does not preserve extended ASCII characters on read

### ISSUE THREAD

**Issue by largo84** [leo-editor/leo-editor#140]
Many extended ASCII characters won't 'round trip' properly using @clean as compared to @file. Simple example:

1) Create new @clean and @file files.
2) Add one or more extended ASCII characters to each file. For example, try ASCII code 246 ÷ (the division symbol, or obelus).
3) Save the file (this creates two new external files).
4) View the external files in another editor (I use Notepad++). The ÷ obelus character shows up in both files as expected.
5) Refresh both from disc (or close the .leo file and reopen).
6) The @clean file now displays a different character (Ã· for ÷), while @file displays ÷ correctly.

Comment: This may be a Python version issue as this behavior does not occur on my Mac OSX installation which uses Python v2.7 (though it could be a Mac encoding difference). On my Windows PC installation, I use Python v3.3.

**Comment by edreamleo** [comment]
On Thu, Feb 26, 2015 at 10:10 AM, Rob Keeney <notifications@github.com>
wrote:

> Many extended ASCII characters won't 'round trip' properly using @clean
> <https://github.com/clean> as compared to @file <https://github.com/file>.
> Simple example:
>
​Thanks for the detailed report.  This will be fixed for b1.

Some preliminary notes:

1. Scite, http://www.scintilla.org/SciTE.html, can detect that the proper
encoding is a code page encoding.  I suppose this isn't too surprising,
since the default scite setting is code.page=0, whatever that means :-)

2. The example file has no BOM, so even chardet, https://chardet.github.io/,
reports the encoding should be utf-8!

3. There seems to be a problem with @encoding cp1252, which seemingly ought
to work on my machine, since locale.getpreferredencoding() reports 'cp1252'
on my machine.  But perhaps your machine has a different code page??

4. The call to codecs.lookup(encoding) in g.isValidEncoding threw a
UnicodeEncodeError(!!), which prevented Leo from loading. That's been
fixed, but I'll have to investigate this kerfuffle further.

5. Using @file instead of @clean does indeed produce *different* results,
though not (on my machine) *correct* results.  This means that simply using
the same (python) code for @clean and @file will not completely solve the
problem.

All in all, an interesting problem.  The workaround, which I recommend in
general, if possible, is to use utf-8 encoding in your original file :-)

Edward

**Comment by largo84** [comment]
Using Notepad++, I set the encoding to UTF-8 and write the following:
Divide by = ÷

Then when I refresh from disk in Leo to @clean, I get:
ï»¿Divide by = Ã·

If I add the following to my .leo @clean node:
@encoding utf-8
Divide by = ÷

Then save and open in Notepad++, I get:
Divide by = รท

Also, Notepad++ doesn't show what type of encoding is used.

**Comment by largo84** [comment]
Another possibly useful clue; this seems to be a recent change. I've had some external files referenced as @nosent for quite a while, but only in the last week or so, did I see where Leo was changing the characters and I would see the changes in @recoverednodes trees. Hope that helps.

**Comment by edreamleo** [comment]
There are bugs here, but they are in the @file logic, not the @clean logic!  Here is my test file on Windows 7:
-----
@encoding cp1252
246 ÷
-----
The non-ascii character is two bytes (c3 b7)  for Unicode utf-8 and a single byte (f7) for cp1252.

This round-trips properly with @clean, without any internal problems.  A file dump shows the file contains hex f7 for the  ÷, for both @clean and @file.

Testing revealed problems with the @file read logic: the encoding specified in the @+leo sentinel line is not being honored. This must be fixed, but it is a separate issue.

In short, it appears that @clean is *more* solid than @file.  The problem you report can be fixed by specifying the proper encoding, most probably cp1252. You can find out for sure by executing this script::

    import locale
    print(locale.getpreferredencoding())

HTH
Edward

**Comment by largo84** [comment]
I ran the script and cp1252 showed up in the terminal window.

If the problem is w/ @file, then why does @clean not 'round trip' the character properly (unless I manually add @encoding cp1252) and @ file does (without needing to add the directive)? I must be missing something.

Does this also mean that I need to remember to add the @encoding directive anytime I need to use extended-ASCII?

**Comment by edreamleo** [comment]
Fixed at rev 89b56b3

In fact, this was a bug in the @file read logic! The fix was to call at.readFileToUnicode in at.readOpenFile.

The new code is enabled by the new_read_line switch in leoAtFile.py.

**Comment by edreamleo** [comment]
@largo84 

> If the problem is w/ @file, then why does @clean not 'round trip' the character properly (unless I manually add @encoding cp1252) and @ file does (without needing to add the directive)?

I'm not sure I was ever seeing the behaviour you describe.  Please test with rev 89b56b3. Trying to explain the internals of Leo is not likely to be helpful.

> Does this also mean that I need to remember to add the @encoding directive anytime I need to use extended-ASCII?

It simply means that the bug has been fixed :-)

You *must* tell Leo what encoding you are using whenever you use a character ch with ord(ch) > 127. That's not a defect of Leo, it's a fact of life.  The most *explicit* way of indicating the encoding is with an @encoding directive.  However, several settings also apply.  See leoSettings.leo, the node::

    @settings-->Unicode options

In particular, @string default_derived_file_encoding = utf-8 Here are the comments in that node::

    The encoding used for derived files if no @encoding directive is in effect.
    This setting is also used to encode files created by the Tangle commands.
    Default is UTF-8 (case not important).

So setting this setting to cp1252 should help.

Another avenue, which I recommend if possible, is to use utf-8 encoding where possible in your *other* tools.  That may be most convenient in the long run.

HTH.  Please feel free to ask question on https://groups.google.com/forum/#!forum/leo-editor.  Unicode issues are of wide interest, and I'm always happy to discuss them.  But here is not the best place.

**Comment by largo84** [comment]
Just pulled the latest revision and have tried out several scenarios (various combinations of @encoding settings and @file and @clean). @file no longer works correctly (it did before). @clean works, at least for ÷. However, it doesn't work for Σ (which doesn't appear to exist in cp1252). It seems that @encode utf-8 is the way forward as it should cover all of the characters I expect to use, but I can't get it to work for either @file or @clean.

**Comment by edreamleo** [comment]
On Mon, Mar 2, 2015 at 10:24 AM, Rob Keeney <notifications@github.com>
wrote:

> It seems that @encode <https://github.com/encode> utf-8 is the way
> forward as it should cover all of the characters I expect to use, but I
> can't get it to work for either @file <https://github.com/file> or @clean
> <https://github.com/clean>.
>
​You shouldn't need to use an @encode directive if you use the default
settings in leoSettings.leo, the node::

    @settings-->Unicode options

These will give you utf-8 in all your external files.

Edward

**Comment by largo84** [comment]
I didn't think so, just making sure. External files appear to write correctly (encoded as UTF-8 without BOM according my editor). They just don't come come back in correctly.

**Comment by edreamleo** [comment]
On Mon, Mar 2, 2015 at 2:50 PM, Rob Keeney <notifications@github.com> wrote:

> I didn't think so, just making sure. External files appear to write
> correctly (encoded as UTF-8 without BOM according my editor). They just
> don't come come back in correctly.
>

​Just double-checked.  It works for me on Windows7.

Edward

**Comment by largo84** [comment]
Hmm, wonder what's different. It doesn't work for me on either win7 or win8 PC (both on Pyton 3.3), but it does on Mac (Python 2.7).

**Comment by edreamleo** [comment]
On Mon, Mar 2, 2015 at 11:05 PM, Rob Keeney <notifications@github.com>
wrote:

> Hmm, wonder what's different. It doesn't work for me on either win7 or
> win8 PC (both on Pyton 3.3), but it does on Mac (Python 2.7).
>
​Please send your test files to me at edreamleo@gmail.com.  I'd like to
clear this up.  Also include your myLeoSettings.leo.

Edward

**Comment by tbnorth** [comment]
On Tue, 03 Mar 2015 04:01:57 -0800
"Edward K. Ream" <notifications@github.com> wrote:

> On Mon, Mar 2, 2015 at 11:05 PM, Rob Keeney <notifications@github.com>
> wrote:
> 
> > Hmm, wonder what's different. It doesn't work for me on either win7
> > or win8 PC (both on Pyton 3.3), but it does on Mac (Python 2.7).
> >
> ​Please send your test files to me at edreamleo@gmail.com.  I'd like
> to clear this up.  Also include your myLeoSettings.leo.

I see a py 2 / 3 difference here.

HOME=/tmp python launchLeo.py

@clean /tmp/test_clean.txt
    Here: ÷ Σ
@file /tmp/test_file.txt
    Here: ÷ Σ

save, quit

grep ^ /tmp/test_*.txt

/tmp/test_clean.txt:Here: ÷ Σ
/tmp/test_file.txt:#@+leo-ver=5-thin
/tmp/test_file.txt:#@+node:tbrown.20150303093202.38: * @file /tmp/test_file.txt
/tmp/test_file.txt:Here: ÷ Σ
/tmp/test_file.txt:#@-leo

saving, reloading, external editing, refresh from disk, no change to ÷ Σ

HOME=/tmp python3 launchLeo.py

both ok inintially, but on refresh from disk (no changes anywhere) 
@file shows �� ��.  Actually this seems a bit inconsistent, sometimes
@file seems to show �� �� straight off the bat without any refresh.

If the @clean file is changed and saved in Leo, Σ becomes £ on write.
The same is try of @file

Leo 5.0-final, build 20150228112315, Sat Feb 28 11:23:15 CST 2015
Git repo info: branch = master, commit = 66ef5fb55d0e
Python 3.4.0, PyQt version 5.2.1
linux

Shell env. settings
LANG=en_US.UTF-8
LANGUAGE=

python -c 'import sys; print(sys.getdefaultencoding())'
ascii

python3 -c 'import sys; print(sys.getdefaultencoding())'
utf-8

HTH :-/

Cheers -Terry

**Comment by largo84** [comment]
That looks like what I'm seeing. Another oddity is that when I use an older version of Leo (v4.11 build 5806 from before I started using Git), I don't see any problems with read/write either ÷ or Σ with @file or @nosent (but still with Python 3.3).

**Comment by edreamleo** [comment]
On Thu, Feb 26, 2015 at 9:20 PM, Rob Keeney <notifications@github.com>
wrote:

> Another possibly useful clue; this seems to be a recent change. I've had
> some external files referenced as @nosent <https://github.com/nosent> for
> quite a while, but only in the last week or so, did I see where Leo was
> changing the characters and I would see the changes in @recoverednodes
> trees. Hope that helps.
>
​First, my apologies.  You stated at the outset that you were using Python
3, yet I ran my tests with Python 2.

On Friday, February 27 rev 89b56b3 changed the @file read logic.​ It may
have affected @clean as well.

This bug is a release blocker.  I'd like to fix it today.

Edward

**Comment by edreamleo** [comment]
On Tue, Mar 3, 2015 at 9:50 AM, tbnorth <notifications@github.com> wrote:

I see a py 2 / 3 difference here.
​ [Fails with 3]

Thanks, Terry.  ​Very helpful.  Now I'll know to run tests with 3 :-)

EKR

**Comment by edreamleo** [comment]
Fixed at rev b198451. All unit tests pass with both Python 2 and 3. Rob's tests files now work as expected.
This was a  bug in the @clean read logic. The fix was a rewrite of at.read_at_nosent_lines.
Note to myself: make sure to run unit tests with Python 3 after any file-related changes.
Please reopen if even the slightest problem remains.

**Comment by largo84** [comment]
Looks good so far; both @clean and @file read/write both ÷ and Σ as expected. I'll check next against some bigger lists and see if there are other issues. Thanks, this really helps me a lot!
